.github/workflows/ci.yml

```diff
@@ -6,13 +6,50 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: terraform fmt
+      - name: cache terraform plugins
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-terraform-plugins
+        with:
+          path: ~/.terraform/plugin-cache
+          key:
+            ${{ runner.os }}-${{ env.cache-name }}-v1-${{ hashFiles('terraform/deployments/**/main.tf') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ env.cache-name }}-v1-${{ hashFiles('terraform/deployments/**/main.tf') }}
+            ${{ runner.os }}-${{ env.cache-name }}-v1-
```

Comment:
If I've understood the docs right, this means the cache keys will contain a hash of the contents of the main.tf's, but on lookup it'll still use a cache from a previous commit if there isn't one with a matching hash. That sounds fine except that I'm not seeing how it'll ever update. The caches get GCed after a week of not being accessed, but that's unlikely to be the case except maybe over the holidays. They don't get GCed based on age, as far as I can tell from the docs.

So in ordinary use, I think this will work as intended until new plugin versions start appearing, at which point the cache will become increasingly ineffective as jobs will continue to use an old cache which won't contain newer plugins, until almost nothing in the cache is relevant any more or until we go a whole week without pushing a commit or raising a PR. That's assuming that Terraform is smart enough to download things which it finds missing from its cache directory (and I have a sneaky suspicion that it's not(!) so maybe it'll even just break once new versions start appearing).

If I've got all that right (probably a big if!), we might want to consider using a datestamp instead of the hash (like in [this example from the docs](https://github.com/actions/cache#creating-a-cache-key)).

Reply:
Yes, I think that's accurate.

At the moment, our [main.tf](https://github.com/alphagov/govuk-infrastructure/blob/main/terraform/deployments/govuk-test/main.tf#L12) files specify a version constraint on the provider (like `version = "~> 3.13"`), so the theory was that we'd keep using the cached one until we chose to upgrade to a newer constraint.

It's not really an issue if it's using a slightly out-of-date provider from its cache when it runs `terraform init` or `terraform validate` - neither of these commands really exercise the plugins.

That said, the provider only takes a couple of seconds to download from GitHub's network, so caching them here is maybe unnecessary complexity. The thing that's important is that they share a cache between deployments, so we don't have to download them _n_ times.

I'd be happy to remove this extra cache if you'd prefer?

Reply:
It seems a bit of a shame not to preserve the cache between builds, since it'd mean a fair bit of unnecessary traffic and wasted resources. We could just give this a try and see if it breaks. If it were me I'd probably use a datestamp instead of the hash though, because I don't think the hash is actually doing anything useful and that way if it goes wrong at least it'll sort itself out after a day.

The [TF docs](https://www.terraform.io/docs/commands/cli-config.html#provider-plugin-cache) say that it'll download the latest version if it doesn't find it in TF_PLUGIN_CACHE_DIR — it won't use an outdated version.

Reply:
Ah, I see. It picks a version before it looks at the cache. I thought it would always use the cached version if that one met the constraints. But you're right, the docs say otherwise.

Reply:
Thanks for spotting this, top reviewing!

I've addressed it in 5053e5c

Reply:
So I think we now effectively have separate TF_PLUGIN_CACHE_DIR per changeset (and also per day). That seems fine. Thanks!